src/saveload/saveload.cpp

```diff
@@ -1620,6 +1639,135 @@ void SlObject(void *object, const SaveLoadTable &slt)
 	}
 }
 
+/**
+ * Save or Load a table header.
+ * Note: a table-header can never contain more than 65535 fields.
+ * @param slt The SaveLoad table with objects to save/load.
+ * @return When loading, the ordered SaveLoad array to use; otherwise an empty list.
+ */
+std::vector<SaveLoad> SlTableHeader(const SaveLoadTable &slt)
+{
+	switch (_sl.action) {
+		case SLA_LOAD_CHECK:
+		case SLA_LOAD: {
+			std::vector<SaveLoad> saveloads;
+
+			/* Build a key lookup mapping based on the available fields. */
+			std::map<std::string, const SaveLoad *> key_lookup;
+			for (auto &sld : slt) {
+				key_lookup[sld.name] = &sld;
+			}
+
+			while (true) {
+				static std::string key;
+				SlObjectMember(nullptr, SLEG_SSTR(key, SLE_STR));
+
+				/* Empty field means end-of-header. */
+				if (key == "") break;
+
+				static uint8 var_type;
+				SlObjectMember(nullptr, SLEG_VAR(var_type, SLE_UINT8));
```

Comment:
I wonder whether I prefer storing the type before the name, so that the sentinel can be a "type" instead of "".